llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -3596,7 +3596,9 @@ Instruction *InstCombinerImpl::visitReturnInst(ReturnInst &RI) {
     if (F->hasRetAttribute(Attribute::NonNull) ||
         (F->getAttributes().getRetDereferenceableBytes() > 0 &&
          !NullPointerIsDefined(F, RetTy->getPointerAddressSpace()))) {
-      if (Value *V = simplifyNonNullOperand(RetVal))
+      if (Value *V = simplifyNonNullOperand(
+              RetVal, /*HasDereferenceable=*/F->getAttributes()
+                              .getRetDereferenceableBytes() > 0))
```

Comment:
Avoid repeated query for dereferenceable bytes? (Should also do this for the arg case.)